crates/ra_assists/src/assists/add_custom_impl.rs

```diff
@@ -0,0 +1,144 @@
+//! FIXME: write short doc here
+
+use hir::db::HirDatabase;
+use ra_syntax::{
+    ast::{self, AstNode},
+    SyntaxKind::{IDENT, WHITESPACE}
+};
+use text_unit::{TextUnit, TextRange};
+use join_to_string::join;
+use smol_str::SmolStr;
+use crate::{Assist, AssistCtx, AssistId};
+
+const DERIVE_TRAIT: &'static str = "derive";
+
+pub(crate) fn add_custom_impl(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let input = ctx.node_at_offset::<ast::AttrInput>()?;
+    let attr = input.syntax().parent().and_then(ast::Attr::cast)?;
+
+    let attr_name = attr.syntax().descendants_with_tokens()
+        .filter(|t| t.kind() == IDENT)
+        .find_map(|i| i.into_token())?
+        .text()
+        .clone();
+
+    if attr_name != DERIVE_TRAIT {
+        return None;
+    }
+
+    let trait_token = ctx.token_at_offset().next().filter(|t| t.kind() == IDENT)?;
+    let annotated = attr.syntax().next_sibling()?;
+    let annotated_name = annotated.text().to_string();
+    let start_offset  = annotated.parent()?.text_range().end();
+
+    ctx.add_action(AssistId("add_custom_impl"), "add custom impl", |edit| {
+        edit.target(attr.syntax().text_range());
+
+        let new_attr_input = input.syntax().descendants_with_tokens()
+            .filter(|t| t.kind() == IDENT)
+            .filter_map(|t| t.into_token().map(|t| t.text().clone()))
+            .filter(|t| t != trait_token.text())
+            .collect::<Vec<SmolStr>>();
```

Comment:
I had to collect here to be able to also get the number of elements.

Reply:
That's OK! The code inside the closure is invoked only when the user actually applies the assist (as opposed to the code outside which is invoked after every cursor movement), so it's OK to be not super efficient here!

Reply:
Nice. I moved it inside the closure with that in mind.